design-proposals/backend-storage-clone-on-migration.md

```diff
@@ -0,0 +1,82 @@
+# Overview
+This proposal is about supporting the migration of all VMs that use a persistent feature, such as persistent vTPM.
+
+To make this document as simple as possible, let's consider that storage in Kubernetes can either be filesystem (FS) or
+block mode, and either ReadWriteOnce (RWO) or ReadWriteMany (RWX) access mode, with storage classes supporting any
+number of combinations of both.
+
+As of KubeVirt 1.3, Here is how each of the 4 combinations is supported by the backend-storage for persistent VM state:
+- RWO FS: supported by backend-storage but makes VMs non-migratable. **This entire design proposal is about removing that limitation.**
+- RWX FS: preferred combination, supported by backend-storage and compatible with VM migrations
+- RWO block: not supported yet, but PRs exist to implement it in a non-migratable way 
+- RWX block: same as above, with migratability functional but considered unsafe, because it leads to 2 pods (source and target virt-launchers) mounting the same ext3 filesystem concurrently.
+
+The above is the best we can do with the current implementation, which involves mounting the same partition on migration source and target.  
+This proposal is about switching to a different model which would instead create a new backend-storage PVC on migration, enabling RWO FS.  
+
+Furthermore, the implementation of this design proposal would also allow a potential block-based implementation, compatible with both RWO and RWX and safely migratable.  
+That, however, is out-of-scope for this design proposal, and may never be needed, since most/all block storage classes also support RWO FS.
+
+## Alternative approach
+The alternative to this design is for upstream components (libvirt - qemu - swtpm) to add block storage support for TPM/EFI.  
+KubeVirt could then create one backend-storage PVC per feature (EFI/TPM) and pass them directly to libvirt.  
+That alternative approach would enable RWX block backend-storage (as opposed to RWO FS in this proposal).  
+
+Pros:
+- Less code to write and maintain in KubeVirt
+- No need to duplicate backend-storage PVCs on migrations
```

Comment:
```suggestion
- No need to duplicate backend-storage PVCs on migrations and delete the correct one after migration success/failure
- libvirt gains an important feature
```

Reply:
Please file a public request on libvirt to add this functionality, and include the URL to it here. The simplest way to "kill" this alternative is if libvirt rejects it. I would think that this would be a mistake on their behalf, but that's their question to answer.

Reply:
I discussed this with a few members of the libvirt team, and they actually don't appear to have strong objections against it.
There would be 4 parts to this:
- block storage support in swtpm: already done
- libvirt glue for swtpm block storage: draft PR open (https://gitlab.com/libvirt/libvirt/-/merge_requests/380)
- block storage support in qemu for NVRAM: could work out of the box, since the NVRAM is already just 1 file (just have to ensure they don't run file-specific operations on it)
- libvirt glue for NVRAM block storage: could also work out of the box, although unlikely, but shouldn't require too much work (just ensuring the "file" is never moved/deleted/chmoded...)

So really not to much left to do overall, however we would have to wait for the first release of every component that needs changes, which could take a long time.
This approach is also definitely not my first choice, and I would hate to request something, have upstream devs work hard on it and ultimately not consume it...
That's why I didn't file any request yet.

Reply:
> This approach is also definitely not my first choice

Please make sure to spell out all your reasoning as its "cons". For example, I did not see the issue of timing written in the proposal.

> That's why I didn't file any request yet.

Let have a formal request and link to it here. We should be explicit that because the reasons spelled out here we may not use it, or may not use it immediately. But let us give the libvirt team an opportunity to refute our assumptions - maybe they can deliver this functionality fast?

Reply:
> Please file a public request on libvirt to add this functionality, and include the URL to it here.

https://issues.redhat.com/browse/RHEL-54325
https://issues.redhat.com/browse/RHEL-54326

Reply:
thanks. github comments can easily get lost. Please add these to the design doc - the fact that they are not done is a reason against the alternative option.

Reply:
Yes, I will rework this design proposal this week and include those links.